File: relay/internal/ntfy/publisher.go
package ntfy

import (
	"context"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
)

// Message is a notification payload for ntfy.
type Message struct {
	Title    string
	Body     string
	Priority string
	Tags     []string
	ClickURL string
}

// Publisher posts messages to an ntfy server topic.
type Publisher struct {
	baseURL string
	topic   string
	token   string
	client  *http.Client
}

func NewPublisher(baseURL, topic, token string, client *http.Client) *Publisher {
	if client == nil {
		client = http.DefaultClient
	}
	return &Publisher{
		baseURL: strings.TrimRight(baseURL, "/"),
		topic:   topic,
		token:   token,
		client:  client,
	}
}

func (p *Publisher) Publish(ctx context.Context, msg Message) error {
	endpoint, err := url.JoinPath(p.baseURL, p.topic)
	if err != nil {
		return fmt.Errorf("ntfy url: %w", err)
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, endpoint, strings.NewReader(msg.Body))
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", "text/plain; charset=utf-8")
	if msg.Title != "" {
		req.Header.Set("Title", msg.Title)
	}
	if msg.Priority != "" {
		req.Header.Set("Priority", msg.Priority)
	}
	if len(msg.Tags) > 0 {
		req.Header.Set("Tags", strings.Join(msg.Tags, ","))
	}
	if msg.ClickURL != "" {
		req.Header.Set("Click", msg.ClickURL)
	}
	if p.token != "" {
		req.Header.Set("Authorization", "Bearer "+p.token)
	}

	res, err := p.client.Do(req)
	if err != nil {
		return err
	}
	defer res.Body.Close()
	body, _ := io.ReadAll(io.LimitReader(res.Body, 4096))
	if res.StatusCode < 200 || res.StatusCode >= 300 {
		return fmt.Errorf("ntfy publish failed: status %d: %s", res.StatusCode, strings.TrimSpace(string(body)))
	}
	return nil
}

File: relay/src/ntfy/publisher.ts
import type { FormattedNtfyNotification } from "./format.js";

export type NtfyConfig = {
  baseUrl: string;
  topic: string;
  accessToken?: string;
};

type FetchFn = typeof fetch;

export class NtfyPublisher {
  constructor(
    private readonly config: NtfyConfig,
    private readonly fetchImpl: FetchFn = fetch,
    private readonly maxAttempts = 3,
  ) {}

  async publish(notification: FormattedNtfyNotification): Promise<void> {
    const url = `${this.config.baseUrl}/${this.config.topic}`;
    const headers: Record<string, string> = {
      Title: notification.title,
      Tags: (notification.tags ?? []).join(","),
    };

    if (notification.click) {
      headers.Click = notification.click;
    }

    if (this.config.accessToken) {
      headers.Authorization = `Bearer ${this.config.accessToken}`;
    }

    let lastError: Error | undefined;

    for (let attempt = 1; attempt <= this.maxAttempts; attempt++) {
      const response = await this.fetchImpl(url, {
        method: "POST",
        headers,
        body: notification.body,
      });

      if (response.ok) {
        return;
      }

      lastError = new Error(`ntfy publish failed with status ${response.status}`);
      if (response.status < 500) {
        throw lastError;
      }
    }

    throw lastError ?? new Error("ntfy publish failed");
  }
}
